src/crossql/Migrations/MigrationRunner.cs

```diff
@@ -20,182 +19,230 @@ public MigrationRunner(IDbProvider dbProvider)
+        public async Task RunAll(SystemRole systemRole, IEnumerable<IDbMigration> migrations)
         {
-            _systemRole = systemRole;
-
             await CreateDatabase().ConfigureAwait(false);
 
-            var orderedMigrations = migrations.OrderBy(m => m.GetType().Name);
-
-            foreach (var migration in orderedMigrations)
-            {
-                var databaseVersion = await GetMigrationInformationAsync(migration).ConfigureAwait(false);
+            var orderedMigrations = migrations.OrderBy(m => m.MigrationVersion);
 
-                await RunBeforeMigration(migration,  databaseVersion).ConfigureAwait(false);
-                await RunMigration(migration,  databaseVersion).ConfigureAwait(false);
-                await RunAfterMigration(migration,  databaseVersion).ConfigureAwait(false);
-            }
+            foreach (var migration in orderedMigrations) await Run(systemRole, migration).ConfigureAwait(false);
         }
 
         public async Task Run(SystemRole systemRole, IDbMigration migration)
         {
-            _systemRole = systemRole;
-
             var databaseVersion = await GetMigrationInformationAsync(migration).ConfigureAwait(false);
 
-            await RunBeforeMigration(migration,  databaseVersion).ConfigureAwait(false);
-            await RunMigration(migration,  databaseVersion).ConfigureAwait(false);
-            await RunAfterMigration(migration,  databaseVersion).ConfigureAwait(false);
+            await RunBeforeMigration(migration, databaseVersion, systemRole).ConfigureAwait(false);
+            await RunMigration(migration, databaseVersion, systemRole).ConfigureAwait(false);
+            await RunAfterMigration(migration, databaseVersion, systemRole).ConfigureAwait(false);
         }
 
-        private async Task RunBeforeMigration(IDbMigration migration, DatabaseVersionModel databaseVersion)
+        private async Task RunBeforeMigration(
+            IDbMigration migration,
+            Version version,
+            SystemRole systemRole)
         {
             // Check Actual DatabaseVersion against the migration version
             // Don't run unless this Migrations BeforeMigration has not been run
-            if (databaseVersion.IsBeforeMigrationComplete == false)
+            if (!version.IsSetupComplete)
             {
-                // Before Migrate
-
-                await migration.RunOrderedMigration(MigrationStep.Setup, _dbProvider).ConfigureAwait(false);
-
-                if (_systemRole == SystemRole.Server)
+                switch (systemRole)
                 {
-                    await migration.RunOrderedMigration(MigrationStep.ServerSetup, _dbProvider).ConfigureAwait(false);
-                }
-                if (_systemRole == SystemRole.Client)
-                {
-                    await migration.RunOrderedMigration(MigrationStep.ClientSetup, _dbProvider).ConfigureAwait(false);
+                    case SystemRole.Server:
```

Comment:
right now it's setup that system specific migrations run BEFORE shared migrations. 
In a perfect world, migrations would be idempotent, but it's just a question... is this the best approach?